src/radio/radio-device-manager.ts

```diff
@@ -0,0 +1,92 @@
+import {RadioController} from './radio-controller';
+import {UdpResponseData} from './dataflow';
+
+/**
+ * The default port to listen on for radio responses.
+ */
+const DEFAULT_LISTEN_PORT = 3311;
+
+/**
+ * An implementation of `smarthome.DeviceManager` that implements radio functionality.
+ */
+export class RadioDeviceManager implements smarthome.DeviceManager {
```

Comment:
Curious if you think it would make sense to move the mocking abstraction one-level below at the controller level and have a common DeviceManager stub for both?

Alternative did you consider folding the controller implement as private method of the manager, or if you want to keep the level of indirection split them away in per-protocol implementation?

Reply:
I think it's worth keeping them separate. I can see two separate use cases:

1. Automated testing using the stubs to place their fulfillment app as the unit under test.  I think in this scenario, we want to restrict the testing context to the handlers themselves as much as possible.  Mocking the radio instead of the `DeviceManager.send()` call would force the developer to think about how the Local Home Platform interprets transport protocol response commands for _every_ unit test involving commands.

2. Manual testing across radio with command line.  In this scenario, the developer doesn't want to mock at all, so merging would not benefit here.  Additionally, **this** is the scenario where it would make sense for the developer to consider how the Local Home Platform interprets their radio responses, since they are doing more integrated testing.

This is just my two cents.

Re: folding `RadioController` into `RadioDeviceManager`, I think it's worth keeping them separate, and with all protocols in one `RadioController`.  It might be worth splitting the protocol implementations into individual classes that are managed within `RadioController`.  In any case I don't think the `RadioDeviceManager` stub is the right place to manage the individual protocol implementations.  I think the `RadioDeviceManager` should just be a middleman that handles translation of handler requests to radio requests, and radio responses to handler responses, with access to a radio API that it can assume works.

Reply:
>  Mocking the radio instead of the DeviceManager.send() call would force the developer to think about how the Local Home Platform interprets transport protocol response commands for every unit test involving commands.

One thing to consider is that with the current implementation we pollute the public interface of `DeviceManager` to enable the unit testing usecase, while we strictly implement the `DeviceManager` for the integration testing usecase.

Currently it seems that the `RadioDeviceManager` does the following:
a. resolve the address from deviceId
b. pass the response port
c. forward UdpRequestData as exploded arguments
d. return UdpResponseData

I'm curious if you think those still be useful in a unit testing context?
- a. could be useful to validate than the id match one that was actually registered
- b. could be a placeholder
- c. and d. could simply forward the command/response to a recording stub.

What we force developer to do really depends on the API that we surface back to the unittest, and in that case it seems that what they would need is:
- checking which *RequestData was passed thru
- specifying which *ResponseData to return

Which could fall within the responsability of the `RadioController` if we adapted its calling convention to consume platform message rather than exploded arguments.

I also wonder if we could simply let the developer specific `sinon.js` [fake](https://sinonjs.org/releases/latest/fakes/) and reuse their expectation interface.

>  I think the RadioDeviceManager should just be a middleman that handles translation of handler requests to radio requests, and radio responses to handler responses, with access to a radio API that it can assume works

This is very sound to isolate the translation layer, but currently it does seems that we are translating twice by introducing an additional interface:
```
DeviceManager *RequestData + resolve device endpoint -> RadioController.sendUdpMessage arguments -> node.js dgram API call
```
Where as we really only need to translate:
```
DeviceManager *RequestData + resolved device endpoint -> node.js dgram API call
```
>  It might be worth splitting the protocol implementations into individual classes that are managed within RadioController

Do you think those could be manager by the DeviceManager? since it will also contain the logic to downcast the command to the right RequestData type.

Reply:
> One thing to consider is that with the current implementation we pollute the public interface of DeviceManager to enable the unit testing usecase, while we strictly implement the DeviceManager for the integration testing usecase.

My perspective is we don't want the developer to speculate about the internals of `RadioController`/`DeviceManager`, since we want to present the class as a blackbox that fulfills radio commands, such that they can test how their handlers react.

I don't know that it would be helpful for the developer to need to mock `*RequestData`, `*ResponseData`.  They would be mocking structs that are used a couple layers down the call stack of `DeviceManager.send()`, when ideally they can ignore those layers and just mock the highest-level call directly.

That being said, I see how requiring the developer to work around the `MockDeviceManager` interface is a similar issue.  I think `sinon.js` might be worth looking into, so I opened issue #53.  It would allow the developer to pick and choose precisely what they want to mock, instead of imposing our interface.

>This is very sound to isolate the translation layer, but currently it does seems that we are translating twice by introducing an additional interface:

My perspective is we do one major translation: LHSDK-domain arguments to a protocol-domain arguments.  The `RadioController` just forwards the translated arguments to a fulfillment class, which plugs them into the respective APIs. 

A couple other things:
- For the chrome version, I'll be able to swap out the `RadioController` for a chrome-specific radio implementation, without re-translating
- After implementing UDP, and TCP, the `RadioDeviceManager` class could be well over 500 lines if implementing everything together, which starts to eat into readability. 

>Do you think those could be manager by the DeviceManager? since it will also contain the logic to downcast the command to the right RequestData type.

This is another option that I will explore if classes become too bloated.  I don't know that we need the flexibility of swapping out protocol-specific implementation classes.

Reply:
> I don't know that it would be helpful for the developer to need to mock *RequestData, *ResponseData

Sorry for the confusion, I wasn't suggestion that developer would need to mock those, but rather that we (or they) could provide a mock `RadioController` that capture them, rather than surfacing dedicated assertions method on the an implementation of the `DeviceManager` platform interface (we could discuss this in #53 if you think that's orthogonal to this change).

>  The RadioController just forwards the translated arguments to a fulfillment class

My understanding was that in its current implementation the `RadioController` was the fulfilment, did you mean `RadioDeviceManager` ?

> I don't know that we need the flexibility of swapping out protocol-specific implementation classes.

My instinct was that swapping them would be useful for testing (setup fake radio, verify it sent and received the right packets) but that could be discussed as part of #53

Reply:
>Sorry for the confusion, I wasn't suggestion that developer would need to mock those, but rather that we (or they) could provide a mock RadioController that capture them, rather than surfacing dedicated assertions method on the an implementation of the DeviceManager platform interface (we could discuss this in #53 if you think that's orthogonal to this change).

I think I understand what you're saying (I'm probably also being a bit unclear with terminology).  My main question was what is the advantage of a `MockRadioController` (or individual radio mock classes.
) vs a `MockDeviceManager`.

>My understanding was that in its current implementation the RadioController was the fulfilment, did you mean RadioDeviceManager ?

Yes, updated my previous comment.

I think the main takeaway is to not reinvent the wheel and consider #53 for mocking.  We can continue discussion in that issue if there's anything else.